tiledb/api/proxy/proxy.h

```diff
@@ -0,0 +1,917 @@
+      lock_state.lock();
+      // Assert: event state is `Hand-off`
+
+      /*
+       * We continue to process events as long as any remain in the queue. In
+       * this phase we process all events. The `will_block` trait only applies
+       * to events when first introduced, not once they've been in the queue.
+       */
+      if (queue_.empty()) {
+        return;
+        /*
+         * Assert: event state is `Change` after the destructor for
+         * `lock_queue` executes.
+         */
+      }
+      event = queue_.front();
+      queue_.pop();
+    }
+  }
+
+ protected:
+  [[nodiscard]] inline ProxyState state() const {
+    return state_;
+  }
+
+  [[nodiscard]] inline std::mutex& m_state() {
+    return m_state_;
+  }
+
+  virtual bool do_construct() = 0;
+  virtual void do_destroy() = 0;
+  virtual void do_shutdown() = 0;
+
+  /**
+   * Default constructor because Proxy needs a default constructor.
+   */
+  ProxyStateMachine()
+      : state_(ProxyState::nascent){};
+
+  void event_access_attach(lock_type& lock) {
+    event(lock, ProxyEvent::access_attach);
+  }
+
+  void event_access_release() {
+    lock_type lock{m_state_, std::defer_lock};
+    event(lock, ProxyEvent::access_release);
+  }
+
+  /**
+   * Predicate whether the state machine is in the `access` state.
+   *
+   * This predicate is used to determine whether to grant access to an
+   * underlying object.
+   */
+  [[nodiscard]] bool is_attached() const {
+    return state_ == ProxyState::access || state_ == ProxyState::last_access;
+  }
+
+ public:
+  /**
+   * Order to construct implemented here for exposure through `Proxy`.
+   *
+   * If validation or construction throws, the proxy will enter its error state.
+   * If validation does not succeed, construction will not happen and the proxy
+   * will stay in the `nascent` state.
+   */
+  void construct() {
+    lock_type lock{m_state_, std::defer_lock};
+    event(lock, ProxyEvent::construct);
+  }
+
+  /**
+   * Order to destroy implemented here for exposure through `Proxy`.
+   */
+  void destroy() {
+    lock_type lock{m_state_, std::defer_lock};
+    event(lock, ProxyEvent::destroy);
+  }
+
+  /**
+   * Order to shut down implemented here for exposure through `Proxy`.
+   */
+  void shutdown() {
+    if constexpr (has_shutdown) {
+      lock_type lock{m_state_, std::defer_lock};
+      event(lock, ProxyEvent::shutdown);
+    } else {
+      destroy();
+    }
+  }
+};
+
+/*
+ * Life cycle concept
+ *
+ * The `LC` template argument of `Proxy` has a number of requirements, some of
+ * which are expressible with C++20 `concept`, some not.
+ *
+ * * The class has a default constructor. This is necessary to allow `Proxy` to
+ *   have a default constructor, since `Proxy` contains a member variable of
+ *   this type.
+ * * The class has a declaration `arguments_type` that holds arguments for the
+ *   constructor during the nascent state of the proxy. The `Proxy` class is
```

Comment:
Will this work if `arguments_type` is variadic?  (I think it probably will but would need to do some experiments.)  That would be a convenient idiom for an LC class to use for building its arguments list.